### PR TITLE
Add module ldap_attrs; deprecate ldap_attr

### LINKED ISSUE #25665: The ldap_attr module fails to initially add/modify olcTLS* attributes
- Bug Report

##### COMPONENT NAME
modules/network/ldap_attr

##### ANSIBLE VERSION
```
ansible 2.3.0.0
  config file =
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Apr  4 2017, 08:47:57) [GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.38)]
```

##### CONFIGURATION
N/A

##### OS / ENVIRONMENT
N/A

##### SUMMARY
OpenLDAP recommends using OLC (cn=config) for configuration rather than the conventional slapd.conf file [1]. Each change via OLC triggers a reload of the OpenLDAP service (slapd) rather than requiring a restart.

Configuring OpenLDAP for TLS/SSL requires at least three attributes in cn=config:

- olcTLSCACertificateFile
- olcTLSCertificateFile
- olcTLSCertificateKeyFile

Initially adding the olcTLSCertificateFile and olcTLSCertificateKeyFile attributes must occur in the same add/modify operation because TLS/SSL requires a server certificate and key. Triggering a reload with only one of these attributes would prevent operation of TLS/SSL, so the OpenLDAP service returns an error (80) if the add/modify operation only contains one of them. The ldap_attr module appears to treat each attribute/value pair as a separate add/modify operation. Thus, attempting to initially add these attributes using the module returns an error and prevents configuration of TLS/SSL. Although the OpenLDAP service returns an error, it still adds one of the two attributes. Running the task a second time usually succeeds because it adds the other attribute which meets the OLC requirement of both attributes present for TLS/SSL. The ldap_attr module should consider these attributes (and possibly other similar situations) as a special case or provide an option to add/modify all attributes in a task using the same add/modify operation.

##### STEPS TO REPRODUCE

- Install OpenLDAP server (with default OLC tree)
- Create a task using ldap_attr to add/modify the olcTLSCertificateFile and olcTLSCertificateKeyFile attributes
- Run task

```yaml
- name: configure openldap server tls/ssl
  ldap_attr:
    dn: "cn=config"
    name: "{{ item.0 }}"
    values: "{{ item.1 }}"
    state: exact
  with_together:
    - [ 'olcTLSCertificateFile', 'olcTLSCertificateKeyFile' ]
    - [ '/etc/ssl/certs/server.pem', '/etc/ssl/private/server.key' ]
```

(Note: The actual task attempts to add/modify all necessary attributes in cn=config using attribute/value pairs from a key/value store.)

##### EXPECTED RESULTS
The ldap_attr module should successfully add/modify the olcTLSCertificateFile and olcTLSCertificateKeyFile attributes.

##### ACTUAL RESULTS
The task fails because the ldap_attr module attempts to add/modify each attribute in a separate add/modify operation.

```
TASK [openldap-server : configure openldap server tls/ssl] ****************************************************************************************************
<1.2.3.4> (0, '\r\n{"msg": "Attribute action failed.", "failed": true, "details": "{\'desc\': \'Other (e.g., implementation specific) error\'}", "invocation": {"module_args": {"dn": "cn=config", "name": "olcTLSCertificateFile", "server_uri": "ldapi:///", "start_tls": false, "bind_dn": null, "state": "exact", "params": null, "bind_pw": "", "values": "/etc/ssl/certs/server.pem"}}}\r\n', 'OpenSSH_7.4p1, LibreSSL 2.5.0\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 18481\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 0\r\nShared connection to 1.2.3.4 closed.\r\n')
failed: [1.2.3.4] (item=[u'olcTLSCertificateFile', u'/etc/ssl/certs/server.pem']) => {
    "details": "{'desc': 'Other (e.g., implementation specific) error'}",
    "failed": true,
    "invocation": {
        "module_args": {
            "bind_dn": null,
            "bind_pw": "",
            "dn": "cn=config",
            "name": "olcTLSCertificateFile",
            "params": null,
            "server_uri": "ldapi:///",
            "start_tls": false,
            "state": "exact",
            "values": "/etc/ssl/certs/server.pem"
        }
    },
    "item": [
        "olcTLSCertificateFile",
        "/etc/ssl/certs/server.pem"
    ],
    "msg": "Attribute action failed."
}
```

[1] https://www.openldap.org/doc/admin24/slapdconf2.html

### PR DESCRIPTION
**Pull request by drybjed**
##### SUMMARY

The current "ldap_attr" module does not support ldap attribute insertations or deletions with objectClass dependencies. The new `ldap_attrs` module uses the same configuration parameter scheme as the `ldap_entry` module and allows easy management of the LDAP entries using Ansible.

This pull request is based on https://github.com/ansible/ansible/pull/31664

Fixes #25665

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

ldap_attrs

##### ADDITIONAL INFORMATION
The new module is currently included in the [DebOps](https://github.com/debops/debops/) project and is used in production environments. Usage examples: [the 'debops.ldap' role tasks](https://github.com/debops/debops/blob/master/ansible/roles/debops.ldap/tasks/ldap_tasks.yml) and [the default configuration data](https://docs.debops.org/en/master/ansible/roles/debops.ldap/defaults/main.html#envvar-ldap__default_tasks), [the `debops.slapd` role tasks](https://github.com/debops/debops/blob/master/ansible/roles/debops.slapd/tasks/slapd_tasks.yml) and [their configuration data](https://docs.debops.org/en/master/ansible/roles/debops.slapd/defaults/main.html#envvar-slapd__default_tasks), including the `X-ORDERED` cn=config attributes.

### CODE REVIEW

File: lib/ansible/modules/net_tools/ldap/_ldap_attr.py
Comment (felixfontein) [inline lib/ansible/modules/net_tools/ldap/_ldap_attr.py]:
```suggestion
  alternative: 'Use M(ldap_attrs) instead. Deprecated in 2.10.'
```
Mentioning `ldap_attrs` once is enough :)

File: lib/ansible/modules/net_tools/ldap/_ldap_attr.py
Comment (felixfontein) [inline lib/ansible/modules/net_tools/ldap/_ldap_attr.py]:
```suggestion
  why: 'The current "ldap_attr" module does not support LDAP attribute insertations or deletions with objectClass dependencies.'
```

File: lib/ansible/modules/net_tools/ldap/ldap_attrs.py
Comment (felixfontein) [inline lib/ansible/modules/net_tools/ldap/ldap_attrs.py:45]: You probably should add yourself to the authors list as well.

File: lib/ansible/modules/net_tools/ldap/ldap_attrs.py
Comment (felixfontein) [inline lib/ansible/modules/net_tools/ldap/ldap_attrs.py]: What's `value`? Do you mean `attributes`?

File: lib/ansible/modules/net_tools/ldap/ldap_attrs.py
Comment (felixfontein) [inline lib/ansible/modules/net_tools/ldap/ldap_attrs.py]: Can you explain what this option does and why it is needed?

From the implementation, it looks like this can be used to overwrite arbitrary module options, without any type checks or other sanity checks. I don't think modules should have such options.

Comment (drybjed) [reply lib/ansible/modules/net_tools/ldap/ldap_attrs.py]: This parameter can be used to [define connection options to the LDAP directory separate from the task itself](https://docs.debops.org/en/master/ansible/roles/debops.ldap/defaults/main.html#envvar-ldap__admin_auth_params). Since different connection methods might need different parameters, it's hard to implement this in a read-only playbook such as DebOps without using a workaround like this. It also gives the users a specific inventory variable to modify if their connection method is different than the default one. The parameter itself comes from the original implementation and is used in the `ldap_entry` module as well.

Perhaps you could reconsider implementing something like this officially for specific Ansible modules, so that the type and sanity checks could be provided? I reckon that many modules that interact with external resources (databases, devices, etc.) with different authentication methods could benefit from this.

Comment (felixfontein) [reply lib/ansible/modules/net_tools/ldap/ldap_attrs.py]: Ansible provides two alternatives for this:
 1. (the more tedious one) provide for every such module option a variable and specify it to every LDAP task in the role via `option_name: "{{ option_name | default(omit) }}"`. Then users can define the options they want, and only these will be passed on to all LDAP tasks. You can potentially save some complexity by using [some YAML tricks](https://docs.ansible.com/ansible/latest/user_guide/playbooks_advanced_syntax.html#yaml-anchors-and-aliases-sharing-variable-values).
 2. use [module defaults](https://docs.ansible.com/ansible/latest/user_guide/playbooks_module_defaults.html). Adding a [module defaults group](https://docs.ansible.com/ansible/latest/user_guide/playbooks_module_defaults.html#module-defaults-groups) for LDAP modules would simplify this (if there are multiple LDAP modules which are used together).

**Comment by felixfontein** [comment]
It looks like you made the module files executable (0755 instead of 0644); this is flagged by the failing sanity tests:
```
lib/ansible/modules/net_tools/ldap/_ldap_attr.py:0:0: module should not be executable
lib/ansible/modules/net_tools/ldap/ldap_attrs.py:0:0: module should not be executable
```

**Comment by drybjed** [comment]
@felixfontein Thanks, force-pushed a fixed version. I need to get rid of that function in `vim`, it's too helpful sometimes.

File: lib/ansible/modules/net_tools/ldap/ldap_attrs.py
Comment (drybjed) [reply lib/ansible/modules/net_tools/ldap/ldap_attrs.py]: YAML anchors are probably more "native" way to do this, although I'm not sure how they work across the Ansible inventory, role default variables and role dependent variables. I would need to research that.

The module defaults might not be a good solution in niche cases, for example managing the OpenLDAP server `cn=config` database requires connecting via the `root` account, and when managing the normal LDAP directory, you would want to bind using a normal admin account over TLS. In such case the module defaults would probably clash. And from the looks of it, this is a playbook-level feature - for my use case it would have to work at least via the Ansible inventory since the playbooks are also meant to be read-only. Defining module defaults via role default variables looks to be not feasible either.

Still, since the `params` option is already present in the `ldap_entry` module, I hope it's not a blocker for merging the `ldap_attrs` module? The functionality could be removed from both modules later if a consensus is reached that it a problematic feature to have in Ansible core. Although it's currently perfect in DebOps case where playbooks and roles are read-only, so it would be sad to see it go.

Edit: It looks like the module defaults might be set on a task level, so that might be a native way to implement `params` functionality. I just need to check if they can use dictionary defined in role default variables and modification via the inventory works. If it does, that's great - we can then remove the `params` from both LDAP modules and introduce the `ldap` module group.

Comment (felixfontein) [reply lib/ansible/modules/net_tools/ldap/ldap_attrs.py]: > YAML anchors are probably more "native" way to do this, although I'm not sure how they work across the Ansible inventory, role default variables and role dependent variables. I would need to research that.

They probably work only in one YAML file. So one can use them locally (in one role's task file), but not accross different YAML files.

> The module defaults might not be a good solution in niche cases, for example managing the OpenLDAP server `cn=config` database requires connecting via the `root` account, and when managing the normal LDAP directory, you would want to bind using a normal admin account over TLS. In such case the module defaults would probably clash. And from the looks of it, this is a playbook-level feature - for my use case it would have to work at least via the Ansible inventory since the playbooks are also meant to be read-only. Defining module defaults via role default variables looks to be not feasible either.

That's true. You can still combine the options from variables (which can come from inventory), though, AFAIK. And you can overwrite the module defaults per task with something else.

> Still, since the `params` option is already present in the `ldap_entry` module, I hope it's not a blocker for merging the `ldap_attrs` module?

I won't merge this PR because of it, but that doesn't mean others are as picky as me :) It would probably be best to discuss this in the Ansible project meeting, to see if such an option would be acceptable for the core team. (If I know it is, I would also merge this :) )

> The functionality could be removed from both modules later if a consensus is reached that it a problematic feature to have in Ansible core. Although it's currently perfect in DebOps case where playbooks and roles are read-only, so it would be sad to see it go.

Removing features is always a lot harder than adding them, so it's better to discuss this in more detail (with more people) before merging.

The best way (if you don't want to use module defaults) is probably to work with variables and `{{ xxx | default(omit) }}`. That also gives more fine control which module options can be overwritten by inventory, and which not. (And by which variables they can be overwritten in which situation.)

File: lib/ansible/modules/net_tools/ldap/_ldap_attr.py
Comment (samccann) [inline lib/ansible/modules/net_tools/ldap/_ldap_attr.py]:
```suggestion
  why: 'The current "ldap_attr" module does not support LDAP attribute insertations or deletions with objectClass dependencies.'
```

File: lib/ansible/modules/net_tools/ldap/ldap_attrs.py
Comment (samccann) [inline lib/ansible/modules/net_tools/ldap/ldap_attrs.py]:
```suggestion
short_description: Add or remove multiple LDAP attribute values
```

File: lib/ansible/modules/net_tools/ldap/_ldap_attr.py
Comment (felixfontein) [inline lib/ansible/modules/net_tools/ldap/_ldap_attr.py]:
```suggestion
  why: 'The current "ldap_attr" module does not support LDAP attribute insertions or deletions with objectClass dependencies.'
```

Comment (felixfontein) [reply lib/ansible/modules/net_tools/ldap/_ldap_attr.py]: (To be honest, I was'n really sure whether this is correct or not; but since I could find `insertation` in neither Merriam-Webster nor Cambridge Dictionary, I would assume it is not correct.)

**Review by samccann — APPROVED** [review]
docs portion LGTM.

**Comment by felixfontein** [comment]
@drybjed could you please remove the `params` option and rebase to resolve the conflict? Without that option, I think the module is ready for merging.

**Comment by drybjed** [comment]
@felixfontein Done. I think that I removed the correct code... I tested the changes in my environment, everything seems to work fine.

File: docs/docsite/rst/porting_guides/porting_guide_2.10.rst
Comment (felixfontein) [inline docs/docsite/rst/porting_guides/porting_guide_2.10.rst]: Please add `The following modules will be removed in Ansible 2.14. Please update update your playbooks accordingly.` above and put your line below that, and not here. The module is not removed in Ansible 2.10, but in Ansible 2.14.

File: lib/ansible/modules/net_tools/ldap/ldap_attrs.py
Comment (felixfontein) [inline lib/ansible/modules/net_tools/ldap/ldap_attrs.py:53]: You should really have at least one example with `state=absent`.

**Comment by drybjed** [comment]
@felixfontein Done. Fingers crossed.

File: lib/ansible/modules/net_tools/ldap/ldap_attrs.py
Comment (felixfontein) [inline lib/ansible/modules/net_tools/ldap/ldap_attrs.py]: Shouldn't this be
```suggestion
        description: []
```
?

**Comment by felixfontein** [comment]
When testing this module, I noticed a potential problem with the way I've been using `ldap_attr`: I'm looping over name/value pairs, like:
```
- ldap_attr:
    name: "{{ item.foo }}"
    values:
      - "{{ item.bar }}"
  with_items: xxx
```
This cannot be made to work with `ldap_attrs` easily, since it is not possible to template dictionary keys. I.e.
```
- ldap_attr:
    attributes:
      "{{ item.foo }}": 
        - "{{ item.bar }}"
  with_items: xxx
```
won't work. In my case, it was easy to rewrite this in a way that works, but it might be harder for other people if they use constructs like that.

Besides that, it seems to work fine! (I used check-mode and only `state: exact` since that is what I need for my data.)

**Comment by drybjed** [comment]
I suppose that's a casuality of making the `ldap_attrs` configuration use the same paradigms that `ldap_entry` uses. Instead of looping over multiple attributes for one LDAP entry, you can loop over multiple LDAP entries using a single list of items.

[Here's an example playbook I'm using to setup new LDAP directory structure](https://github.com/debops/debops/blob/master/ansible/playbooks/ldap/init-directory.yml#L68) - as you can see it's a list of LDAP entries. The `debops.ldap` role is written in such a way that you can define an LDAP entry multiple times and activate different configuration entries conditionally, changing the outcome. You could use this to generate a list of LDAP objects from some data via a `lookup('template')` and pass that list to [two Ansible tasks](https://github.com/debops/debops/blob/master/ansible/roles/debops.ldap/tasks/ldap_tasks.yml) which are [generated dynamically during playbook execution](https://github.com/debops/debops/blob/master/ansible/roles/debops.ldap/tasks/main.yml#L67-L81) to make whatever LDAP tasks you want. For example, here's [a list of LDAP tasks](https://docs.debops.org/en/master/ansible/roles/debops.ldap/defaults/main.html#envvar-ldap__default_tasks) included by default in the `debops.ldap` role.

I think that this method of managing LDAP directories is far superior than the old `ldap_attr` module which couldn't be mixed with `ldap_entry` module as easily. But in the end, I think that for very large LDAP data sets, it's still much more efficient to generate LDIF files and import them out-of-band on the LDAP server, or use a custom script to mangle the data from one source to the other. Using Ansible to import 3000+ users into LDAP with round-trip Python scripts for each user would take lots of time.

**Comment by felixfontein** [comment]
Yep, definitely. In fact, I had already written two different codes for `ldap_entry` and `ldap_attr`, and now was able to get rid of the `ldap_attr` variant and replace it with the same dict creation as for `ldap_entry`.

**Comment by felixfontein** [comment]
(Looks like I forgot to send off the review comment I had. Here it is.)

File: lib/ansible/modules/net_tools/ldap/ldap_attrs.py
Comment (drybjed) [reply lib/ansible/modules/net_tools/ldap/ldap_attrs.py]: I think that should work, but I test it first to be sure.

Comment (drybjed) [reply lib/ansible/modules/net_tools/ldap/ldap_attrs.py]: I think that will work, but I'll test it first.

File: lib/ansible/modules/net_tools/ldap/ldap_attrs.py
Comment (felixfontein) [inline lib/ansible/modules/net_tools/ldap/ldap_attrs.py]: I don't think this is a good idea. This prevents the valid value `""`.
```suggestion
        else:
```

**Review by felixfontein — APPROVED** [review]
LGTM

**Comment by felixfontein** [comment]
@Noles @drybjed thank you both very much for implementing/fixing this!

**Comment by drybjed** [comment]
@felixfontein Awesome, thanks for merging!

**Pull request merged**
